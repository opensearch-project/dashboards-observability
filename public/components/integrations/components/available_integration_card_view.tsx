/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiPanel,
  EuiCard,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiCompressedFieldSearch,
  EuiButtonGroup,
} from '@elastic/eui';
import React, { useState } from 'react';
import {
  AvailableIntegrationsCardViewProps,
  AvailableIntegrationType,
} from './available_integration_overview_page';
import { INTEGRATIONS_BASE } from '../../../../common/constants/shared';
import { badges } from './integration_category_badge_group';
import { basePathLink } from '../../../../common/utils/shared';

export function AvailableIntegrationsCardView(props: AvailableIntegrationsCardViewProps) {
  const [toggleIconIdSelected, setToggleIconIdSelected] = useState('1');

  const getImage = (url?: string) => {
    let optionalImg;
    if (url) {
      optionalImg = (
        <img style={{ height: 100, width: 100 }} alt="" className="synopsisIcon" src={url} />
      );
    }
    return optionalImg;
  };

  const toggleButtonsIcons = [
    {
      id: '0',
      label: 'list',
      iconType: 'list',
    },
    {
      id: '1',
      label: 'grid',
      iconType: 'grid',
    },
  ];

  const onChangeIcons = (optionId: string) => {
    setToggleIconIdSelected(optionId);
    if (optionId === '0') {
      props.setCardView(false);
    } else {
      props.setCardView(true);
    }
  };

  const renderRows = (integrations: AvailableIntegrationType[]) => {
    if (!integrations || !integrations.length) return null;
    return (
      <>
        <EuiFlexGroup gutterSize="l" style={{ flexWrap: 'wrap' }}>
          {integrations.map((i, v) => {
            return (
              <EuiFlexItem key={v} style={{ minWidth: '14rem', maxWidth: '14rem' }}>
                <EuiCard
                  icon={getImage(
                    basePathLink(
                      `${INTEGRATIONS_BASE}/repository/${i.name}/static/${i.statics.logo.path}`
                    )
                  )}
                  title={i.displayName ? i.displayName : i.name}
                  description={i.description}
                  data-test-subj={`integration_card_${i.name.toLowerCase()}`}
                  titleElement="span"
                  href={basePathLink(`/app/integrations#/available/${i.name}`)}
                  footer={badges(i.labels ?? [])}
                />
              </EuiFlexItem>
            );
          })}
        </EuiFlexGroup>
        <EuiSpacer />
      </>
    );
  };

  return (
    <EuiPanel>
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem>
          <EuiCompressedFieldSearch
            fullWidth
            isClearable={false}
            placeholder="Search..."
            data-test-subj="search-bar-input-box"
            onChange={(e) => {
              props.setQuery(e.target.value);
            }}
          />
        </EuiFlexItem>
        <EuiFlexItem grow={false}>{props.filters}</EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButtonGroup
            legend="Text align"
            options={toggleButtonsIcons}
            idSelected={toggleIconIdSelected}
            onChange={(id) => onChangeIcons(id)}
            isIconOnly
          />
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer />
      {renderRows(props.data.hits.filter((x) => x.name.includes(props.query)))}
    </EuiPanel>
  );
}
