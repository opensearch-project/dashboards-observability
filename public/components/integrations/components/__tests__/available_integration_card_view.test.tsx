/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { waitFor } from '@testing-library/react';
import { AvailableIntegrationsCardView } from '../available_integration_card_view';
import { availableCardViewData } from './testing_constants';
import React from 'react';

describe('Available Integration Card View Test', () => {
  configure({ adapter: new Adapter() });

  it('Renders nginx integration card view using dummy data', async () => {
    const wrapper = mount(<AvailableIntegrationsCardView {...availableCardViewData} />);

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
