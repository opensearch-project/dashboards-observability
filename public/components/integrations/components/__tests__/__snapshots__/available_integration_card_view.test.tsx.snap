// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Available Integration Card View Test Renders nginx integration card view using dummy data 1`] = `
<AvailableIntegrationsCardView
  data={
    Object {
      "hits": Array [
        Object {
          "assets": Array [
            Object {
              "extension": "ndjson",
              "name": "nginx",
              "type": "savedObjectBundle",
              "version": "1.0.1",
            },
          ],
          "author": "John Doe",
          "components": Array [
            Object {
              "name": "communication",
              "version": "1.0.0",
            },
            Object {
              "name": "http",
              "version": "1.0.0",
            },
            Object {
              "name": "logs",
              "version": "1.0.0",
            },
          ],
          "description": "Nginx HTTP server collector",
          "displayName": "NginX Dashboard",
          "license": "Apache-2.0",
          "name": "nginx",
          "sourceUrl": "https://github.com/",
          "statics": Object {
            "gallery": Array [
              Object {
                "annotation": "NginX Dashboard",
                "path": "dashboard1.png",
              },
              Object {
                "annotation": "NginX Logo",
                "path": "logo.svg",
              },
            ],
            "logo": Object {
              "annotation": "NginX Logo",
              "path": "logo.svg",
            },
          },
          "type": "logs",
          "version": "1.0.1",
        },
      ],
    }
  }
  http={[MockFunction]}
  isCardView={false}
  query=""
  renderCateogryFilters={[Function]}
  setCardView={[Function]}
  setQuery={[Function]}
>
  <EuiPanel>
    <div
      className="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    >
      <EuiFlexGroup
        gutterSize="s"
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <EuiFlexItem>
            <div
              className="euiFlexItem"
            >
              <EuiFieldSearch
                compressed={false}
                data-test-subj="search-bar-input-box"
                fullWidth={true}
                incremental={false}
                isClearable={false}
                isLoading={false}
                onChange={[Function]}
                placeholder="Search..."
              >
                <EuiFormControlLayout
                  compressed={false}
                  fullWidth={true}
                  icon="search"
                  isLoading={false}
                >
                  <div
                    className="euiFormControlLayout euiFormControlLayout--fullWidth"
                  >
                    <div
                      className="euiFormControlLayout__childrenWrapper"
                    >
                      <EuiValidatableControl>
                        <input
                          className="euiFieldSearch euiFieldSearch--fullWidth"
                          data-test-subj="search-bar-input-box"
                          onChange={[Function]}
                          onKeyUp={[Function]}
                          placeholder="Search..."
                          type="search"
                        />
                      </EuiValidatableControl>
                      <EuiFormControlLayoutIcons
                        compressed={false}
                        icon="search"
                        isLoading={false}
                      >
                        <div
                          className="euiFormControlLayoutIcons"
                        >
                          <EuiFormControlLayoutCustomIcon
                            size="m"
                            type="search"
                          >
                            <span
                              className="euiFormControlLayoutCustomIcon"
                            >
                              <EuiIcon
                                aria-hidden="true"
                                className="euiFormControlLayoutCustomIcon__icon"
                                size="m"
                                type="search"
                              >
                                <EuiIconBeaker
                                  aria-hidden={true}
                                  className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                  focusable="false"
                                  role="img"
                                  style={null}
                                >
                                  <svg
                                    aria-hidden={true}
                                    className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                    focusable="false"
                                    height={16}
                                    role="img"
                                    style={null}
                                    viewBox="0 0 16 16"
                                    width={16}
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                                    />
                                  </svg>
                                </EuiIconBeaker>
                              </EuiIcon>
                            </span>
                          </EuiFormControlLayoutCustomIcon>
                        </div>
                      </EuiFormControlLayoutIcons>
                    </div>
                  </div>
                </EuiFormControlLayout>
              </EuiFieldSearch>
            </div>
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
          >
            <div
              className="euiFlexItem euiFlexItem--flexGrowZero"
            />
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
          >
            <div
              className="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <EuiButtonGroup
                idSelected="1"
                isIconOnly={true}
                legend="Text align"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "iconType": "list",
                      "id": "0",
                      "label": "list",
                    },
                    Object {
                      "iconType": "grid",
                      "id": "1",
                      "label": "grid",
                    },
                  ]
                }
              >
                <fieldset
                  className="euiButtonGroup euiButtonGroup--small euiButtonGroup--text"
                  disabled={false}
                >
                  <EuiScreenReaderOnly>
                    <legend
                      className="euiScreenReaderOnly"
                    >
                      Text align
                    </legend>
                  </EuiScreenReaderOnly>
                  <div
                    className="euiButtonGroup__buttons"
                  >
                    <EuiButtonGroupButton
                      color="text"
                      element="label"
                      iconType="list"
                      id="0"
                      isDisabled={false}
                      isIconOnly={true}
                      isSelected={false}
                      key="0"
                      label="list"
                      name="random_html_id"
                      onChange={[Function]}
                      size="s"
                    >
                      <EuiButtonDisplay
                        baseClassName="euiButtonGroupButton"
                        className="euiButtonGroupButton-isIconOnly"
                        color="text"
                        element="label"
                        fill={false}
                        htmlFor="random_html_id"
                        iconType="list"
                        isDisabled={false}
                        onClick={[Function]}
                        size="s"
                        textProps={
                          Object {
                            "className": "euiScreenReaderOnly",
                            "data-text": "list",
                            "ref": [Function],
                            "title": "list",
                          }
                        }
                      >
                        <label
                          className="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small euiButtonGroupButton-isIconOnly"
                          disabled={false}
                          htmlFor="random_html_id"
                          onClick={[Function]}
                          style={
                            Object {
                              "minWidth": undefined,
                            }
                          }
                        >
                          <EuiButtonContent
                            className="euiButton__content"
                            iconSide="left"
                            iconType="list"
                            textProps={
                              Object {
                                "className": "euiButton__text euiScreenReaderOnly",
                                "data-text": "list",
                                "ref": [Function],
                                "title": "list",
                              }
                            }
                          >
                            <span
                              className="euiButtonContent euiButton__content"
                            >
                              <EuiIcon
                                className="euiButtonContent__icon"
                                color="inherit"
                                size="m"
                                title="list"
                                type="list"
                              >
                                <EuiIconBeaker
                                  aria-hidden={true}
                                  className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                                  focusable="false"
                                  role="img"
                                  style={null}
                                  title="list"
                                  titleId="random_html_id"
                                >
                                  <svg
                                    aria-hidden={true}
                                    aria-labelledby="random_html_id"
                                    className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                                    focusable="false"
                                    height={16}
                                    role="img"
                                    style={null}
                                    viewBox="0 0 16 16"
                                    width={16}
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <title
                                      id="random_html_id"
                                    >
                                      list
                                    </title>
                                    <path
                                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                                    />
                                  </svg>
                                </EuiIconBeaker>
                              </EuiIcon>
                              <span
                                className="euiButton__text euiScreenReaderOnly"
                                data-text="list"
                                title="list"
                              >
                                <input
                                  checked={false}
                                  className="euiScreenReaderOnly"
                                  data-test-subj="0"
                                  disabled={false}
                                  id="random_html_id"
                                  name="random_html_id"
                                  onChange={[Function]}
                                  type="radio"
                                />
                                list
                              </span>
                            </span>
                          </EuiButtonContent>
                        </label>
                      </EuiButtonDisplay>
                    </EuiButtonGroupButton>
                    <EuiButtonGroupButton
                      color="text"
                      element="label"
                      iconType="grid"
                      id="1"
                      isDisabled={false}
                      isIconOnly={true}
                      isSelected={true}
                      key="1"
                      label="grid"
                      name="random_html_id"
                      onChange={[Function]}
                      size="s"
                    >
                      <EuiButtonDisplay
                        baseClassName="euiButtonGroupButton"
                        className="euiButtonGroupButton-isSelected euiButtonGroupButton-isIconOnly"
                        color="text"
                        element="label"
                        fill={true}
                        htmlFor="random_html_id"
                        iconType="grid"
                        isDisabled={false}
                        onClick={[Function]}
                        size="s"
                        textProps={
                          Object {
                            "className": "euiScreenReaderOnly",
                            "data-text": "grid",
                            "ref": [Function],
                            "title": "grid",
                          }
                        }
                      >
                        <label
                          className="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small euiButtonGroupButton--fill euiButtonGroupButton-isSelected euiButtonGroupButton-isIconOnly"
                          disabled={false}
                          htmlFor="random_html_id"
                          onClick={[Function]}
                          style={
                            Object {
                              "minWidth": undefined,
                            }
                          }
                        >
                          <EuiButtonContent
                            className="euiButton__content"
                            iconSide="left"
                            iconType="grid"
                            textProps={
                              Object {
                                "className": "euiButton__text euiScreenReaderOnly",
                                "data-text": "grid",
                                "ref": [Function],
                                "title": "grid",
                              }
                            }
                          >
                            <span
                              className="euiButtonContent euiButton__content"
                            >
                              <EuiIcon
                                className="euiButtonContent__icon"
                                color="inherit"
                                size="m"
                                title="grid"
                                type="grid"
                              >
                                <EuiIconBeaker
                                  aria-hidden={true}
                                  className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                                  focusable="false"
                                  role="img"
                                  style={null}
                                  title="grid"
                                  titleId="random_html_id"
                                >
                                  <svg
                                    aria-hidden={true}
                                    aria-labelledby="random_html_id"
                                    className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                                    focusable="false"
                                    height={16}
                                    role="img"
                                    style={null}
                                    viewBox="0 0 16 16"
                                    width={16}
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <title
                                      id="random_html_id"
                                    >
                                      grid
                                    </title>
                                    <path
                                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                                    />
                                  </svg>
                                </EuiIconBeaker>
                              </EuiIcon>
                              <span
                                className="euiButton__text euiScreenReaderOnly"
                                data-text="grid"
                                title="grid"
                              >
                                <input
                                  checked={true}
                                  className="euiScreenReaderOnly"
                                  data-test-subj="1"
                                  disabled={false}
                                  id="random_html_id"
                                  name="random_html_id"
                                  onChange={[Function]}
                                  type="radio"
                                />
                                grid
                              </span>
                            </span>
                          </EuiButtonContent>
                        </label>
                      </EuiButtonDisplay>
                    </EuiButtonGroupButton>
                  </div>
                </fieldset>
              </EuiButtonGroup>
            </div>
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
      <EuiSpacer>
        <div
          className="euiSpacer euiSpacer--l"
        />
      </EuiSpacer>
      <EuiFlexGroup
        gutterSize="l"
        style={
          Object {
            "flexWrap": "wrap",
          }
        }
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
          style={
            Object {
              "flexWrap": "wrap",
            }
          }
        >
          <EuiFlexItem
            key="0"
            style={
              Object {
                "maxWidth": "14rem",
                "minWidth": "14rem",
              }
            }
          >
            <div
              className="euiFlexItem"
              style={
                Object {
                  "maxWidth": "14rem",
                  "minWidth": "14rem",
                }
              }
            >
              <EuiCard
                data-test-subj="integration_card_nginx"
                description="Nginx HTTP server collector"
                footer={<EuiBadgeGroup />}
                onClick={[Function]}
                title="NginX Dashboard"
                titleElement="span"
              >
                <EuiPanel
                  className="euiCard euiCard--centerAligned euiCard--isClickable"
                  data-test-subj="integration_card_nginx"
                  element="div"
                  hasShadow={true}
                  onClick={[Function]}
                >
                  <div
                    className="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--isClickable euiCard euiCard--centerAligned euiCard--isClickable"
                    data-test-subj="integration_card_nginx"
                    onClick={[Function]}
                  >
                    <div
                      className="euiCard__content"
                    >
                      <EuiTitle
                        className="euiCard__title"
                        id="random_html_idTitle"
                        size="s"
                      >
                        <span
                          className="euiTitle euiTitle--small euiCard__title"
                          id="random_html_idTitle"
                        >
                          <button
                            aria-describedby=" random_html_idDescription"
                            className="euiCard__titleButton"
                            disabled={false}
                            onClick={[Function]}
                          >
                            NginX Dashboard
                          </button>
                        </span>
                      </EuiTitle>
                      <EuiText
                        className="euiCard__description"
                        id="random_html_idDescription"
                        size="s"
                      >
                        <div
                          className="euiText euiText--small euiCard__description"
                          id="random_html_idDescription"
                        >
                          <p>
                            Nginx HTTP server collector
                          </p>
                        </div>
                      </EuiText>
                    </div>
                    <div
                      className="euiCard__footer"
                    >
                      <EuiBadgeGroup>
                        <div
                          className="euiBadgeGroup euiBadgeGroup--gutterExtraSmall"
                        />
                      </EuiBadgeGroup>
                    </div>
                  </div>
                </EuiPanel>
              </EuiCard>
            </div>
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
      <EuiSpacer>
        <div
          className="euiSpacer euiSpacer--l"
        />
      </EuiSpacer>
    </div>
  </EuiPanel>
</AvailableIntegrationsCardView>
`;
